File: mandi-direct/backend/app/schemas/matching.py
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, Field
from app.models.enums import ConfidenceLevel, MatchLevel


class MatchSignal(BaseModel):
    signal_type: str = Field(..., description="CATEGORY, PRODUCT, VARIETY, QUALITY, LOCATION, QUANTITY, PRICE, PURCHASE_HISTORY")
    signal_name: str
    matched: bool
    points_awarded: float
    max_points: float
    description: str


class MatchResult(BaseModel):
    match_score: float = Field(..., ge=0, le=100, description="Deterministic match score out of 100")
    match_level: MatchLevel
    confidence: ConfidenceLevel
    explanation: str
    matched_signals: List[MatchSignal] = Field(default_factory=list)
    unmatched_signals: List[MatchSignal] = Field(default_factory=list)
    generated_at: datetime


class BuyerSafeInfo(BaseModel):
    buyer_id: str
    display_name: str
    buyer_type: str = "Verified Buyer"
    district: Optional[str] = None
    state: Optional[str] = None
    preferred_categories: List[str] = Field(default_factory=list)
    interested_products: List[str] = Field(default_factory=list)
    verified_purchases_count: int = 0


class FarmerProduceSafeInfo(BaseModel):
    produce_id: str
    product_name: str
    category: str
    variety: Optional[str] = None
    quality_grade: str
    expected_price: float
    price_unit: str
    available_quantity: float
    quantity_unit: str
    primary_image_url: Optional[str] = None
    village: Optional[str] = None
    mandal: Optional[str] = None
    district: str
    state: str
    farmer_id: str
    farmer_name: str
    is_verified_farmer: bool
    demand_level: Optional[str] = None
    demand_trend: Optional[str] = None


class BuyerMatchResult(BaseModel):
    buyer: BuyerSafeInfo
    produce: FarmerProduceSafeInfo
    match: MatchResult


class ProductMatchResult(BaseModel):
    produce: FarmerProduceSafeInfo
    match: MatchResult


class FarmerBuyerMatchesResponse(BaseModel):
    items: List[BuyerMatchResult]
    total: int
    page: int
    page_size: int
    total_pages: int
    high_match_count: int = 0
    generated_at: datetime


class BuyerProductMatchesResponse(BaseModel):
    items: List[ProductMatchResult]
    total: int
    page: int
    page_size: int
    total_pages: int
    generated_at: datetime


class ProduceBuyerMatchesResponse(BaseModel):
    produce: FarmerProduceSafeInfo
    items: List[BuyerMatchResult]
    total: int
    page: int
    page_size: int
    total_pages: int
    high_match_count: int = 0
    generated_at: datetime

File: mandi-direct/frontend/src/types/matching.ts
export type MatchLevel =
  | "VERY_LOW"
  | "LOW"
  | "MODERATE"
  | "HIGH"
  | "VERY_HIGH"
  | "INSUFFICIENT_DATA";

export type ConfidenceLevel = "LOW" | "MEDIUM" | "HIGH";

export interface BuyerPreferences {
  id?: string;
  buyer_user_id?: string;
  preferred_categories: string[];
  preferred_products: string[];
  preferred_varieties: string[];
  preferred_quality_grades: string[];
  preferred_districts: string[];
  preferred_states: string[];
  minimum_quantity: number | null;
  maximum_quantity: number | null;
  minimum_price: number | null;
  maximum_price: number | null;
  created_at?: string;
  updated_at?: string;
}

export interface MatchSignal {
  signal_type: string;
  signal_name: string;
  matched: boolean;
  points_awarded: number;
  max_points: number;
  description: string;
}

export interface MatchResult {
  match_score: number;
  match_level: MatchLevel;
  confidence: ConfidenceLevel;
  explanation: string;
  matched_signals: MatchSignal[];
  unmatched_signals: MatchSignal[];
  generated_at: string;
}

export interface BuyerSafeInfo {
  buyer_id: string;
  display_name: string;
  buyer_type: string;
  district?: string | null;
  state?: string | null;
  preferred_categories: string[];
  interested_products: string[];
  verified_purchases_count: number;
}

export interface FarmerProduceSafeInfo {
  produce_id: string;
  product_name: string;
  category: string;
  variety?: string | null;
  quality_grade: string;
  expected_price: number;
  price_unit: string;
  available_quantity: number;
  quantity_unit: string;
  primary_image_url?: string | null;
  village?: string | null;
  mandal?: string | null;
  district: string;
  state: string;
  farmer_id: string;
  farmer_name: string;
  is_verified_farmer: boolean;
  demand_level?: string | null;
  demand_trend?: string | null;
}

export interface BuyerMatchItem {
  buyer: BuyerSafeInfo;
  produce: FarmerProduceSafeInfo;
  match: MatchResult;
}

export interface ProductMatchItem {
  produce: FarmerProduceSafeInfo;
  match: MatchResult;
}

export interface FarmerBuyerMatchesResponse {
  items: BuyerMatchItem[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
  high_match_count: number;
  generated_at: string;
}

export interface BuyerProductMatchesResponse {
  items: ProductMatchItem[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
  generated_at: string;
}

export interface ProduceBuyerMatchesResponse {
  produce: FarmerProduceSafeInfo;
  items: BuyerMatchItem[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
  high_match_count: number;
  generated_at: string;
}

export interface MatchingFilters {
  produce_id?: string;
  category?: string;
  product_name?: string;
  variety?: string;
  quality_grade?: string;
  district?: string;
  state?: string;
  min_score?: number;
  match_level?: string;
  sort?: string;
  page?: number;
  page_size?: number;
}
